File: algorithms/python/backtracking/travelling_salesman_problem.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""travelling_salesman_problem.py

Rat in a maze is an example to display how backtrackng canbe
used to solve the problem.

.. _PEP 484:
    https://www.python.org/dev/peps/pep-0484/
"""

from sys import maxsize
from itertools import permutations

V = 4

def travelling_salesman_problem(graph: list[list], s: int):
    """Implementation of traveling Salesman Problem.

    Args:
        graph (Any): .
        s (Any): .

    Returns:
        int: Prints out results.
    """ 
    # store all vertex apart from source vertex
    vertex = []
    for i in range(V):
        if i != s:
            vertex.append(i)
 
    # store minimum weight Hamiltonian Cycle
    min_path = maxsize
    next_permutation=permutations(vertex)
    for i in next_permutation:
 
        # store current Path weight(cost)
        current_pathweight = 0
 
        # compute current path weight
        k = s
        for j in i:
            current_pathweight += graph[k][j]
            k = j
        current_pathweight += graph[k][s]
 
        # update minimum
        min_path = min(min_path, current_pathweight)
         
    return min_path

if __name__ == '__main__':
    travelling_salesman_problem()

File: algorithms/backtracking/travelling-salesman-problem/zig/travelling_salesman_problem.zig
const std = @import("std");

var maxInt = i32.max;

fn factorial(n: i32) i32 {
    var result: i32 = 1;
    for (1..(n + 1)) |i| {
        result *= i;
    }
    return result;
}

fn travellingSalesmanProblem(graph: [][]i32, s: i32) i32 {
    const V = graph.len;
    var vertex = []i32{};

    for (graph) |row, i| {
        if (i != s) {
            vertex |= i;
        }
    }

    var minPath: i32 = maxInt;
    var permutations = std.math.factorial(vertex.len);

    for (0..permutations) |permutation| {
        var currentPathWeight: i32 = 0;
        var k: i32 = s;

        for (vertex) |j| {
            currentPathWeight += graph[k][j];
            k = j;
        }
        currentPathWeight += graph[k][s];

        minPath = @intCast(i32, @min(@intCast(u32, minPath), @intCast(u32, currentPathWeight)));
    }

    return minPath;
}

pub fn main() void {
    const V = 4;
    var graph = [][]i32{
        []i32{0, 10, 15, 20},
        []i32{10, 0, 35, 25},
        []i32{15, 35, 0, 30},
        []i32{20, 25, 30, 0},
    };

    const s = 0;
    const minPath = travellingSalesmanProblem(graph, s);

    if (minPath == maxInt) {
        std.debug.print("-1\n", .{});
    } else {
        std.debug.print("Minimum cost of TSP: {}\n", .{minPath});
    }
}
